### PR TITLE
[Flex_counter]: Remove poll interval from db key

### PR DESCRIPTION
**Pull request by sihuihan88**
Signed-off-by: Sihui Han <sihan@microsoft.com>

### CODE REVIEW

**Comment by sihuihan88** [comment]
depends on https://github.com/Azure/sonic-swss-common/pull/182

**Comment by sihuihan88** [comment]
retest this please

File: syncd/syncd.cpp [hunk 1]
```diff
@@ -3292,14 +3260,14 @@ int syncd_main(int argc, char **argv)
...
-    std::shared_ptr<swss::DBConnector> dbFlexCounter = std::make_shared<swss::DBConnector>(PFC_WD_DB, swss::DBConnector::DEFAULT_UNIXSOCKET, 0);
+    std::shared_ptr<swss::DBConnector> dbFlexCounter = std::make_shared<swss::DBConnector>(FLEX_COUNTER_DB, swss::DBConnector::DEFAULT_UNIXSOCKET, 0);
 
     g_redisClient = std::make_shared<swss::RedisClient>(dbAsic.get());
 
     std::shared_ptr<swss::ConsumerTable> asicState = std::make_shared<swss::ConsumerTable>(dbAsic.get(), ASIC_STATE_TABLE);
     std::shared_ptr<swss::NotificationConsumer> restartQuery = std::make_shared<swss::NotificationConsumer>(dbAsic.get(), "RESTARTQUERY");
-    std::shared_ptr<swss::ConsumerStateTable> flexCounterState = std::make_shared<swss::ConsumerStateTable>(dbFlexCounter.get(), PFC_WD_STATE_TABLE);
-    std::shared_ptr<swss::ConsumerStateTable> flexCounterPlugin = std::make_shared<swss::ConsumerStateTable>(dbFlexCounter.get(), PLUGIN_TABLE);
+    std::shared_ptr<swss::ConsumerTable> pfcWdFlexCounter = std::make_shared<swss::ConsumerTable>(dbFlexCounter.get(), FLEX_PFC_WD_TABLE);
```
Comment (marian-pritsak) [inline syncd/syncd.cpp]: Flex counter was designed in a way that it's agnostic to its clients. In the case when we'll need to implement, for example, another watchdog, flex counter's code will need to be modified. I don't think that removing polling interval from key with a cost of introducing this dependency is a good thing to do.

**Comment by sihuihan88** [comment]
retest this please

**Comment by sihuihan88** [comment]
retest this please

**Comment by sihuihan88** [comment]
@marian-pritsak Could you help to review? Thanks :)

**Comment by lguohan** [comment]
>                 SWSS_LOG_ERROR("Object type not supported");

minor comments: better to print out the objectType and the field to give more debug information.

---
Refers to: syncd/syncd.cpp:2682 in 103a7f8. [](commit_id = 103a7f853b8e4fb966e6b4148bf73087c495a57b, deletion_comment = False)

File: syncd/syncd_flex_counter.cpp [hunk 2]
```diff
@@ -30,6 +30,17 @@ FlexCounter::QueueAttrIds::QueueAttrIds(
...
 
+void FlexCounter::setPollInterval(
+        _In_ uint32_t pollInterval,
+        _In_ std::string instanceId)
+{
+    SWSS_LOG_ENTER();
+
+    FlexCounter &fc = getInstance(instanceId);
+    fc.m_pollInterval = pollInterval;
+    // fc.startFlexCounterThread();
```
Comment (lguohan) [inline syncd/syncd_flex_counter.cpp]: >   // fc.startFlexCounterThread(); [](start = 1, length = 34)

can we remove this?

File: syncd/syncd_flex_counter.h [hunk 3]
```diff
@@ -82,9 +87,9 @@ class FlexCounter
             std::vector<sai_port_stat_t> portCounterIds;
         };
 
-        FlexCounter(uint32_t pollInterval);
-        static FlexCounter& getInstance(uint32_t pollInterval);
-        static void removeInstance(uint32_t pollInterval);
+        FlexCounter(std::string instanceId);
```
Comment (lguohan) [inline syncd/syncd_flex_counter.h]: >instanceId [](start = 32, length = 10)

can you explain this instanceId concept? I am not familiar with it.

Comment (lguohan) [reply syncd/syncd_flex_counter.h]: It looks like the instanceId is the flex counter group name, so maybe change to groupname is more clear?

---
In reply to: [166061744](https://github.com/Azure/sonic-sairedis/pull/289#discussion_r166061744) [](ancestors = 166061744)

Comment (sihuihan88) [reply syncd/syncd_flex_counter.h]: Each group will be an instance in flex counter and instance id is the group name.

File: syncd/syncd_flex_counter.cpp [hunk 4]
```diff
@@ -208,53 +219,58 @@ void FlexCounter::removePort(
...
+    FlexCounter &fc = getInstance(instanceId);
 
     auto counterIter = fc.m_queueCounterIdsMap.find(queueVid);
-    if (counterIter != fc.m_queueCounterIdsMap.end())
+    if (counterIter == fc.m_queueCounterIdsMap.end())
     {
-        fc.m_queueCounterIdsMap.erase(counterIter);
-        found = true;
+        SWSS_LOG_ERROR("Trying to remove nonexisting queue counter Ids 0x%lx", queueVid);
+        if (fc.m_queueCounterIdsMap.empty() && fc.m_portCounterIdsMap.empty() && fc.m_queueAttrIdsMap.empty())
```
Comment (lguohan) [inline syncd/syncd_flex_counter.cpp]: >fc.m_queueCounterIdsMap.empty() && fc.m_portCounterIdsMap.empty() && fc.m_queueAttrIdsMap.empty( [](start = 12, length = 96)

lots of duplicated code to check empty counter id, suggest to have a function to check it.

Comment (lguohan) [reply syncd/syncd_flex_counter.cpp]: for example 

IsCounterIdsMapEmpty()

---
In reply to: [166065518](https://github.com/Azure/sonic-sairedis/pull/289#discussion_r166065518) [](ancestors = 166065518)

File: syncd/syncd_flex_counter.cpp [hunk 5]
```diff
@@ -208,53 +219,58 @@ void FlexCounter::removePort(
     // Remove flex counter if counter IDs map is empty
     if (fc.m_queueCounterIdsMap.empty() && fc.m_portCounterIdsMap.empty() && fc.m_queueAttrIdsMap.empty())
     {
-        removeInstance(pollInterval);
+        removeInstance(instanceId);
     }
 }
 
 void FlexCounter::removeQueue(
```
Comment (lguohan) [inline syncd/syncd_flex_counter.cpp]: for removeQueue, I think you changed the logic.

In marian's code, he tries to remove the queue id both from m_queueCounterIdsMap and m_queueAttrIdsMap. 

In your code, if the first one is not found, you will return and will not remove the queueid from the second map.

Why do you change the logic here? Is this intended? IMO, marian code is more correct.

File: syncd/syncd.cpp [hunk 6]
```diff
@@ -2535,41 +2535,83 @@ sai_status_t processEvent(
...
+            {
+                auto shaStrings = swss::tokenize(value, ',');
+                for (const auto &sha : shaStrings)
+                {
+                    FlexCounter::addPortCounterPlugin(sha, key);
+                }
+            }
+            else
+            {
+                SWSS_LOG_ERROR("Object type not supported");
```
Comment (lguohan) [inline syncd/syncd.cpp]: >SWSS_LOG_ERROR("Object type not supported"); [](start = 16, length = 44)

should be field not supported, also please print out the actual field for debugging purpose.

File: syncd/syncd.cpp [hunk 7]
```diff
@@ -2535,41 +2535,83 @@ sai_status_t processEvent(
...
+        _In_ swss::ConsumerTable &consumer)
 {
     SWSS_LOG_ENTER();
 
     std::lock_guard<std::mutex> lock(g_mutex);
-
     swss::KeyOpFieldsValuesTuple kco;
     consumer.pop(kco);
 
     const auto &key = kfvKey(kco);
```
Comment (lguohan) [inline syncd/syncd.cpp]: >key [](start = 16, length = 3)

-> groupname

File: syncd/syncd_flex_counter.cpp (hunk 5 above)
Comment (sihuihan88) [reply syncd/syncd_flex_counter.cpp]: Marian's new logic was added recently and I didn't rebase it before force push. Will update and sorry for the inconvenience

**Pull request merged**